### PR TITLE
Support multiple formats in LiveView

### PR DESCRIPTION
**Pull request by bcardarella**
This PR introduces formats to LiveView. Now LV can render multiple content type formats not just HTML. This is to allow LiveView Native to treat the platforms (i.e. `swiftui`, `jetpack`, `winui3`) as formats instead of just using query params.

This PR will also allow for format specific layouts to be rendered in LiveView instead of always `html`. In LiveView Native we had to introduce a bunch of macro work-arounds to get layout rendering to work. This will do away with at least half of our code to leverage the internal format rendering this PR allows for.

There are some breaking changes and I will highlight each of those below. But in most cases this should be a backwards compatible change. For example, now for `use Phoenix.LiveView` it should get `format`, and `layouts` just like `Phoenix.Controller` does:

```elixir
# both of these configurations are necessary

# this first will allow the `HTML.Engine` to be configured to allow for a format-specific tag_handler
config :phoenix_live_view, :tag_handlers, [
  gameboy: LiveViewNative.TagEngine,
]

# this will allow the new formats to declare their engine. Because we can override the tag_handler
# I expect this to almost always be `Phoenix.HTML.Engine` and we could make it a default if no overriding engine is provided
config :phoenix_template, :format_encoders, [
  gameboy: Phoenix.HTML.Engine
]

def MyAppWeb.Layout do
  use MyAppWeb, :html

  embed_templates "layouts/*html"
  embed_templates "layouts/*gameboy", suffix: "_gameboy"
end

def MyAppWeb.HomeLive do
  use Phoenix.LiveView,
    formats: [:html, :gameboy],
    layouts: [
      html: {MyAppWeb.Layout, :app},
      gameboy: {MyAppWeb.Layout, :app_gameboy}
    ]
end
```

However the prior form still works:

```elixir
def MyAppWeb.HomeLive do
  use Phoenix.LiveView, layout: {MyAppWeb.Layout, :app}
end
```

In this case `:formats` defaults to `[:html]` and the `:layout` key will be dropped and replaced with `layouts: [html: {MyAppWeb.Layout, :app}]`

I did not introduce multiple formats to function component or LiveView components yet as I want to first get feedback on this implementation as the PR is growing large.

When providing the `live_layout` in your `mount` function I did not implement any format specificity here. Because we have `render_with` I believe that branching on formats should delegate to `render_with` and within those render functions should then return the desired layout if it is to be overridden:

```elixir
def MyAppWeb.HomeLive do
  def mount(params, session, socket) do
    case get_format(session) do
      :html -> {:ok, socket, layout: {MyAppWeb.Layouts, :custom_layout}}
      :gameboy -> render_with(socket, &MyAppWeb.GameboyLive.render/1, layout: MyAppWeb.Layouts, :custom_gameboy)
    end
  end
end
```

Because root layouts are only rendered for DeadViews there really wasn't anything to change in LV itself as long as the formats are set up properly for DeadView rendering in the Router

TODO:

- [ ] update documentation
- [ ] write tests for non-html formats

### CODE REVIEW

**Comment by bcardarella** [comment]
I was going to tag this as a `Draft` but that doesn't appear to be an option for me.

File: lib/phoenix_live_view.ex [hunk 1]
```diff
@@ -343,12 +344,19 @@ defmodule Phoenix.LiveView do
   defmacro __using__(opts) do
     # Expand layout if possible to avoid compile-time dependencies
     opts =
-      with true <- Keyword.keyword?(opts),
-           {layout, template} <- Keyword.get(opts, :layout) do
-        layout = Macro.expand(layout, %{__CALLER__ | function: {:__live__, 0}})
-        Keyword.replace!(opts, :layout, {layout, template})
+      if Macro.quoted_literal?(opts) do
```
Comment (bcardarella) [inline lib/phoenix_live_view.ex:347]: I * believe* this is compatible with the prior implementation for expanding the aliases.

File: lib/phoenix_live_view/html_engine.ex [hunk 2]
```diff
@@ -30,11 +30,27 @@ defmodule Phoenix.LiveView.HTMLEngine do
       trim: trim,
       caller: __CALLER__,
       source: source,
-      tag_handler: __MODULE__,
+      tag_handler: get_tag_handler(path),
```
Comment (josevalim) [inline lib/phoenix_live_view/html_engine.ex:33]: Instead of changing the tag handler here, you should provide your own engines that pass their own tag handler.

**Comment by josevalim** [comment]
Hi @bcardarella!

Honestly, I am not sure why we need to go in this direction, since we can do the content negotiation and set the render funtion plus layout based on the on_mount callback. Something like this:

```elixir
use Phoenix.LVN, formats: [...]
```

which will expand to:

```
defmacro __using__(_opts) do
  quote do
    on_mount {Phoenix.LVN, {:content_negotiation, unquote(opts)}}
  end
end
```

where

```elixir
def on_mount({:content_negotiation, opts}, ...) do
  # return {:cont, render_with(socket, ...), layout: ...}
end
```

I am not sure if we support setting the layout from `on_mount`, but if we don't, we can either add that or add `put_layout` function. My point is that we do not need to add the whole content negotiation machinary to LiveView, we should rather provide the necessary building blocks.

I know one of the points is being similar with Phoenix.Controllers but controllers and liveview have one big difference in that controllers allows you to process the connection way before content negotiation happens, so we need to expose more, LV/LVN do not need to do that, so we can do it upfront.

---

So to sum up, let's look at where the complexity is in LVN and provide small helpers to remove the complexity there. In particular, I can think about two places for improvement:

1. Make it easier to define your own engines, as we don't want to publicly expose the tag handler configuration (i.e. you should define an engine per format). If you have links to your own engine definition, I can take a look and provide feedback

2. Make it easier to set the layout from `on_mount`, if not yet possible

**Comment by bcardarella** [comment]
@josevalim Chris was adament that we not write our own engine when we prposed that a while back. This PR doesn't just set the `app` layout it also does the format implicit template rendering when no `render/1` function is present. So `home_live.swiftui.heex` templates can be rendered.

**Comment by josevalim** [comment]
Chris was adamant to not *write* your own engine. You are not writing your own, you are using the TagEngine which was designed to work with LVN. My point is not to write your own, but to define your own engine that calls TagEngine with the proper parameters, instead of hardcoding format logic inside the official HTML engine. :)

**Comment by bcardarella** [comment]
@josevalim we proposed replacing HTMLEngine with something of our own we were told not to do that.

**Comment by bcardarella** [comment]
@josevalim if what you're suggesting is that all of the non-HTML parts of HTMLEngine are put into another module so they can be `use`d to create a new Engine without having to write everything over again and we can define our own tag handler I'm :+1: on that

**Comment by josevalim** [comment]
> This PR doesn't just set the app layout it also does the format implicit template rendering when no render/1 function is present.

That's also an area I am concerned about. I don't think you can have multiple colocated templates in the same LiveView, because each engine will use different helpers. Mixing them all in the same module is asking for trouble, because you will try to call `.form` and render a HTML form in the middle of your SwiftUI. And even if you namespace them, importing the functions from 2 or 3 different engines into the same view is asking for trouble.

**Comment by josevalim** [comment]
> @josevalim if what you're suggesting is that all of the non-HTML parts of HTMLEngine are put into another module so they can be used to create a new Engine without having to write everything over again and we can define our own tag handler I'm 👍 on that

Yes, that's what I am proposing. But I believe the annotate `:debug_annotation`and `:annotate_tag_handler` APIs are currently private. So remove those bits and if you need to use them in the future, let us know. :)

File: lib/phoenix_live_view/controller.ex [hunk 3]
```diff
@@ -41,7 +43,7 @@ defmodule Phoenix.LiveView.Controller do
         conn
         |> Phoenix.Controller.put_view(LiveView.Static)
         |> Phoenix.Controller.render(
-          "template.html",
+          "template." <> get_format(conn),
```
Comment (josevalim) [inline lib/phoenix_live_view/controller.ex:46]: I have pushed a fix for this on main. :)

**Comment by bcardarella** [comment]
> Mixing them all in the same module is asking for trouble

I agree when it comes to enforcing isolation, but `render_with` while allowing us to delegate to a function that *could* be in another module isn't *guaranteed* to be in another module. So we have the possibility that devs could choose a local function rather than within module isolation.

> HTMLEngine

I can give this a go, but I'd like to get @chrismccord's sign off on this direction first before I put time into it

---

This PR doesn't really address components either. To your prior point of mixed helpers / isolation I am of the mind that function components shouldn't have mixing of content handling either, and should be isolated entirely. Where it gets a bit grey for me is in LV components, one value prop we've seen with LVN is with event handlers, sharing the event handling and state management between `html`, `swiftui`, `jetpack` clients in a single place has a big impact on app development time. We're handling state once, and rendering out the UI per-client. Re-writing LV components for a specific client each time would impact app development time, rather than allowing for per-client templates or having something similiar to `render_with` for LV components would be nice.

**Comment by josevalim** [comment]
> This PR doesn't just set the app layout it also does the format implicit template rendering when no render/1 function is present.

What we can do is to support a single format. So if you put `live_view.html.heex` or `live_view.native.swiftui`, then it will use whatever is available. However, you can't have both (because as I said, I think it is a bad idea). If you need both, then you must use separate views. Does that work for you?

**Comment by josevalim** [comment]
> I can give this a go, but I'd like to get @chrismccord's sign off on this direction first before I put time into it

You don't need Chris sign-off. Chris delegated this particular area to me, so we are fine. :)

**Comment by josevalim** [comment]
To make sure we are on the same page, the template is 100% allowed, expected to work, and already discussed with @chrismccord:


```
defmodule Phoenix.LiveView.SwiftUIEngine do
  @moduledoc """
  The SwiftUIEngine that powers `.swiftui` templates and the `~SUI` sigil.

  It works by adding a native parsing and validation layer on top
  of `Phoenix.LiveView.TagEngine`.
  """

  @behaviour Phoenix.Template.Engine

  @impl true
  def compile(path, _name) do
    # We need access for the caller, so we return a call to a macro.
    quote do
      require Phoenix.LiveView.SwiftUIEngine
      Phoenix.LiveView.SwiftUIEngine.compile(unquote(path))
    end
  end

  @doc false
  defmacro compile(path) do
    source = File.read!(path)

    EEx.compile_string(source,
      engine: Phoenix.LiveView.TagEngine,
      line: 1,
      file: path,
      trim: true,
      caller: __CALLER__,
      source: source,
      tag_handler: __MODULE__
    )
  end

  # Now implement the TagEngine as you wish
  @behaviour Phoenix.LiveView.TagEngine
end
```

**Comment by bcardarella** [comment]
> What we can do is to support a single format

So I don't think a single format would be realistic as the odds that someone is only going to build for just a single client be it swiftui, jetpack, winui3, then also only limit themselves to a single target be it mobile, tablet, tv, watch, or other I think is low. Enforcing module isolation for rendering I think is the way to go. Leaving event handling and state management to the LV.

**Comment by josevalim** [comment]
> So we have the possibility that devs could choose a local function rather than within module isolation.

It is totally fine if they choose a separate function and use embed_templates to embed that. My point is that `render/1` should not magically perform content negotiation based on collocated templates. `render/1` either renders a single unambiguous template, or none. :)

**Comment by bcardarella** [comment]
> should not magically perform content negotiation based on collocated templates

100%, tbh it felt nasty writing that

**Comment by josevalim** [comment]
@bcardarella hrm... I guess the case to support different formats in implicit render is iOS/tvOS? do they use the same engine/helpers and everything?

**Comment by bcardarella** [comment]
@josevalim there are certain views and modifiers that are target specific, at least in SwiftUI. While you do get some responsive-design-like things like the same UI can run on phone and tablet you'll want a different UI for watch and for macos. We haven't fully explored VisionOS yet but I imagine this may be closer to MacOS app building

**Comment by bcardarella** [comment]
LVN blows up the rendering target complexity by ∞% 😮‍💨

**Comment by josevalim** [comment]
Ok. So let me know how you want to proceed. I think there are a couple options here:

1. Allow the implicit render format to be configured (in case you know your LVN has a single format)
2. Allow the implicit render format to be disabled (so you don't bother looking for a HTML template for LVN cases)
3. Keep it as is

But I will keep the content negotiation off the table. :)

**Comment by josevalim** [comment]
> LVN blows up the rendering target complexity by ∞% 😮‍💨

No worries, my goal is to enable you to support all cases, but without necessarily adding the logic to LV. :) I am confident we will find the proper hooks and callbacks.

**Comment by bcardarella** [comment]
If we can guarantee module isoltion I'd prefer to defer to the rendering module. That way I can write something like: 

```elixir
defmodule MyAppWeb.HomeLive.SwiftUI do
  use LiveVIewNative.Renderer
end
```

and in `lib/live/home_live/swiftui/` the template would live if no `render/1` function is defined.

The other issue with `render_with/2` here is that because the function being delegated to can be anything we don't have a way to guarantee a check if that function exists or not. Someone could easily delegate to `SwiftUI.other_render/1` but in `LiveViewNative.Renderer` at compile-time I need to check if `render/1` exists or not for the implicit template logic.

**Comment by bcardarella** [comment]
Closing with some notes. Per a conversation with @josevalim the path forward:

1. extend `on_mount/1` to accept the `:layout` option by extending the tuple to take three elements #2963 
2. Write `LiveViewNative.NativeEngine` or we may even need an engine per-client, unsure yet: https://github.com/phoenixframework/phoenix_live_view/pull/2962#issuecomment-1868281235

DeadView Renders work just fine with formats via the ACCEPT header for `root_layout`

Jose provided content negotiation for Static LV https://github.com/phoenixframework/phoenix_live_view/commit/84d07a633cabc1de2b83752b9c4aa75b2ca48136
